librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
                                   + dskew[idx[:, 0], idx[:, 1]])
 
     return pitches, mags
+
+
+def yin(y, sr=22050, win_length=1024, hop_length=None, fmin=55, fmax=900, peak_threshold=0.1):
+    '''Fundamental frequency (F0) estimation. [1]_
+
+    .. [1] De Cheveigné, A., & Kawahara, H. (2002).
+        "YIN, a fundamental frequency estimator for speech and music."
+        The Journal of the Acoustical Society of America, 111(4), 1917-1930.
+
+     Parameters
+     ----------
+     y : np.ndarray [shape=(n,)]
+         audio time series.
+
+     sr : number > 0 [scalar]
+         sampling rate of `y` in Hertz.
+
+     win_length : int [scalar]
+         length of the window in samples.
+         By default, win_length=1024 corresponds to a time scale of 46 ms at
+         a sampling rate of 22050 Hz.
+
+     hop_length : None or int > 0 [scalar]
+         number of audio samples between adjacent pYIN predictions.
+         If `None`, defaults to `frame_length // 4`.
+
+     fmin: number > 0 [scalar]
+         minimum frequency in Hertz.
+
+     fmax: number > 0 [scalar]
+         maximum frequency in Hertz.
+
+     peak_threshold: number > 0 [scalar]
+         absolute threshold for peak estimation.
+
+     Returns
+     -------
+     f0: np.ndarray [shape=(n_frames,)]
+         time series of fundamental frequencies in Hertz.
+
+     Examples
+     --------
+     Computing a fundamental frequency curve from an audio input:
+     >>> y = librosa.chirp(440, 880, duration=7.0)
+     >>> yin(y)
+     array([441.15076669, 441.6640896 , 442.16219657, ...,
+        871.59470743, 872.63699131, 873.66696896])
+     '''
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = int(win_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Autocorrelation.
+    y_frames = util.frame(y, frame_length=2*win_length, hop_length=hop_length)
+    n_frames = y_frames.shape[-1]
+    a = np.fft.fft(y_frames, 2*win_length, axis=0)
+    b = np.fft.fft(y_frames[win_length::-1, :], 2*win_length, axis=0)
+    acf_frames = np.fft.ifft(a*b, 2*win_length, axis=0).real[win_length:]
```

Comment:
Coming back on autocorrelation...

Let's think through what's going on here.  Since each frame is processed independently here, we may as well think of a single buffer of `N = 2 * win_length` samples.

`a` is the FFT of the buffer.

`b` is the FFT of the first half of the buffer, reversed, and zero-padded out to `N` frequency samples again.  (This is functionally what setting `n>len(x)` in FFT does.)

If we want to implement this in the frequency domain, the truncate-and-pad could be achieved by convolution against a sinc (in the time domain, it's equivalently a pointwise multiply with a box).  Reversing only the first half of the signal is a little funny, and explains why this isn't the same as `a.conj()`.  If anything, it should be something like `a.conj() * exp(-pi * j * k)`  (for k'th DFT bin, since we're effectively reversing the windowed frame, and then circular-shifting by N/2 samples) and convolved with a sinc.  So even-numbered DFT bins have no change and odd DFT bins are sign-reversed. Wild.  This seems to more or less check out numerically (more or less because the shift could be N/2 +- 1, so it's not quite as clean as previously mentioned).

Looking qualitatively at the two calculations, they seem pretty similar:
```python
win_length = 1024
hop_length = win_length // 4

yf = librosa.util.frame(y, 2*win_length, hop_length=hop_length)

ac1 = librosa.autocorrelate(yf, win_length, axis=0)

a = np.fft.rfft(yf, axis=0)
b = np.fft.rfft(yf[win_length::-1, :], n=yf.shape[0], axis=0)
ac2 = np.fft.irfft(a * b, axis=0)[win_length:]
```
![image](https://user-images.githubusercontent.com/1190540/83805199-6d1d9700-a67d-11ea-8769-13b58e9966bb.png)

Picking individual frames out, they generally have the same peak profiles, but the envelope can be different:
frame 5:
![image](https://user-images.githubusercontent.com/1190540/83807557-48c3b980-a681-11ea-997e-278f54972422.png)

frame 10:
![image](https://user-images.githubusercontent.com/1190540/83807575-4f523100-a681-11ea-81dc-7431782da711.png)


frame 13:
![image](https://user-images.githubusercontent.com/1190540/83807586-54af7b80-a681-11ea-8103-c5ff2c319d04.png)

Note: these are using `win_length` frames for `autocorrelate` and `2*win_length` frames for the direct method.

The autocorrelate implementation always decays with lag, which makes sense to me.  It works by taking the frame, padding it out to double+1 length, and then inverting its power spectrum; so higher lags rely on more zero-padding.  The direct implementation here only zero-pads one side of the calculation, so the decay isn't as severe if the signal is stationary.

This is all to say: I think I have a handle on what's going on now, and I guess it's okay to leave it as implemented (though it should use rfft instead of fft :)).  I have a hunch that we could boost performance a little bit by putting a window function in the `b` calculation, eg 

``python
b = np.fft.rfft(yf[win_length::-1, :] * librosa.filters.get_window('hann', win_length+1)[:, np.newaxis], n=yf.shape[0], axis=0)
``

but it's a very minor thing.

Reply:
Exactly, the YIN paper uses this modified autocorrelation function to avoid the tapering.

From the paper:
> The second problem is that the sum involves more terms for small [lags] than for large. This introduces an unwanted bias that can be corrected by dividing each sample by [a tapering window size].  However, it remains that large [lag] samples are derived from a smaller window of data, and are thus less stable than small [lag] samples.

I'll go ahead and replace the fft by rfft though.

Reply:
> Exactly, the YIN paper uses this modified autocorrelation function to avoid the tapering.

Sorry, thanks for bearing with me as I reason it out from first principles. :)

I wonder if it's worth generalizing our autocorrelate implementation to support this?  It seems generally useful, also in tempogram calculations.  But I'll leave that as a separate issue.

Reply:
@bmcfee this part of the code is originally from me. I agree it is not efficient, but that's the best i could do given the boundary constraints.
IIRC the original YIN code is based on cumulative sums and (gasp) repmats

https://github.com/kylerbrown/yin/blob/master/yin/yin2.m#L48

Reply:
the difficulty in adapting the MATLAB yin code to librosa is that the original code has a hop_length of 1.

But @bmcfee you're totally right, having used fftconvolve was silly. We could use a difference of cumsum's, which would result in linear (as opposed to N log T) complexity.